### PR TITLE
BGE-M3 Performance Optimization for Batch 1->32 and ISL 512

### PR DESCRIPTION
**Pull request by gtobarTT**
### PR Category

Performance

### Summary

This PR brings the single-card BGE-M3 optimization work into the Galaxy (`bge_m3_2cq`) branch and rounds out the pooling / test / docs story.

**Performance (single card, Blackhole P150, S=512, bf8_b):**

| Batch | Before | After | Speedup | PCC |
|------:|-------:|------:|:-------:|:---:|
| B8    | 48.98 ms | 24.30 ms | **2.02×** | 0.944 |
| B16   | 94.76 ms | 38.57 ms | **2.46×** | 0.947 |

Key wins: fidelity gates (LoFi/HiFi2 + fp32-dest off where safe), tuned MLP-wi 2D mcast program configs, fused QKV/concat-heads custom ops, L1 placement for MLP-wi/wo + attention output, and SDPA tuning. The biggest B16 win is full-bf8 SDPA (both `score_dtype` and `qkv_dtype` -> bf8, halving Q/K/V bandwidth). Every change is PCC-gated at the repo standard **0.94**.

**Functionality:**
- Adds `pooling="sparse"` as a first-class mode to `BgeM3Model` (alongside `None`/`cls`/`mean`/`colbert`). The model emits the **raw head output**; downstream post-processing (crop, masking, vocab scatter, scoring) is the caller's responsibility, matching how the vLLM `BgeM3ForEmbedding` wrapper consumes the heads.
- Adds a `--pooling {none,cls,mean,colbert}` flag to the DP=32 multi-process harness (`dp_multiprocess.py`).
- Adds `ttnn.copy_device_to_torch`: a fused D2H op that reads a device tensor straight into a pre-allocated `torch.Tensor`, removing the intermediate ttnn host-staging tensor (one fewer host-to-host memcpy on the D2H path). Used by the perf/DP harnesses; documented in `api.rst`.

**Tests:**
- New `test_model_pooling.py` — mirrors `test_generator_vllm.py`'s 4 cases (dense/cls, colbert, sparse, sparse corner-case) but drives the core model via `create_tt_model(pooling=...)` instead of the vLLM wrapper. Uses the **same fixed reference tensors and Blackhole tolerances** as the vLLM test.

**Galaxy throughput (P150, S=512, e2e latency, median):**

| DP | Batch/chip | Global batch | Total e2e (ms) | Emb/s | Tok/s |
|---:|-----------:|-------------:|---------------:|------:|------:|
| 1  | 1  | 1    | 4.60  | 217.4  | 0.11M |
| 32 | 1  | 32   | 4.73  | 6,417  | 3.29M |
| 32 | 8  | 256  | 29.12 | 8,470  | 4.34M |
| 32 | 16 | 512  | 49.42 | 10,014 | 5.13M |
| 32 | 32 | 1024 | 79.96 | 12,274 | 6.28M |

(Throughput is measured on the full H2D -> Forward -> D2H pipeline, slowest-chip gated — not forward-only. Per-chip latency is within ~1% of DP=1, confirming clean independent-chip scaling.)

**Docs / cleanup:**
- `USER_GUIDE.md` -> `README.md`, with a new **Pooling methods** section documenting all `pooling=` modes, their output shapes (`B`/`S`/`D` legend), and the head-weight loading note.

- Added `ttnn.copy_device_to_torch`: a fused D2H op that reads a device tensor straight into a pre-allocated `torch.Tensor`, removing the intermediate ttnn host-staging tensor (one fewer
 host-to-host memcpy on the D2H path). Used by the perf/DP harnesses; documented in `api.rst`
 
## CI Run

- [x] [(Single-card) Demo tests (no perf)](https://github.com/tenstorrent/tt-metal/actions/runs/27958397971)
- [ ] [(Single-card) Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/runs/27958433822)

### CODE REVIEW

File: models/demos/wormhole/bge_m3/tt/custom_ops/fused_concat_heads/kernels/reader_concat_heads_batched.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    constexpr uint32_t cb_id = 0;
+    constexpr uint32_t block_tiles = in0_c * in0_w_tiles;
+
+    const uint32_t tile_size_bytes = get_tile_size(cb_id);
+    const auto s0 = TensorAccessor(in0_args, in0_tensor_addr);
+
+    for (uint32_t block = 0; block < num_blocks; block++) {
+        // Read the whole concat block (num_heads * head_dim_tiles) in one go.
+        cb_reserve_back(cb_id, block_tiles);
```
Comment (bbradelTT) [inline models/demos/wormhole/bge_m3/tt/custom_ops/fused_concat_heads/kernels/reader_concat_heads_batched.cpp]: Please use device 1.1/2.0 APIs. See https://github.com/tenstorrent/tt-metal/blob/main/docs/source/tt-metalium/tt_metal/apis/kernel_apis/data_movement/device_api_migration_guide.md

**Review by bbradelTT — COMMENTED** [review]
Please comment on or address AI comments.

**Review by arginugaTT — APPROVED** [review]

File: models/demos/wormhole/bge_m3/tt/custom_ops/fused_concat_heads/kernels/reader_concat_heads_headsplit.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    for (uint32_t w = 0; w < num_work_units; ++w) {
+        const uint32_t work_unit = work_unit_start + w;
+        const uint32_t block = work_unit / head_groups;
+        const uint32_t group = work_unit - block * head_groups;
+        const uint32_t batch = block / in0_h_tiles;
+        const uint32_t h_tile = block - batch * in0_h_tiles;
+        const uint32_t head_start = group * heads_per_group;
+
+        cb_reserve_back(cb_id, group_tiles);
```
Comment (bbradelTT) [inline models/demos/wormhole/bge_m3/tt/custom_ops/fused_concat_heads/kernels/reader_concat_heads_headsplit.cpp]: Please use device 2.0 APIs.
You can point AI to https://github.com/tenstorrent/tt-metal/blob/main/docs/source/tt-metalium/tt_metal/apis/kernel_apis/data_movement/device_api_migration_guide.md and existing examples.

Please do a check via something like `grep -e 'cb_.*(' -e 'noc_.*(' -r *` and make sure that's all clean for the files you're changing.

Comment (gtobarTT) [reply models/demos/wormhole/bge_m3/tt/custom_ops/fused_concat_heads/kernels/reader_concat_heads_headsplit.cpp]: thank, will do

**Review by bbradelTT — APPROVED** [review]
Please add a note about copy_device_to_torch to the description.

File: ttnn/cpp/ttnn-nanobind/pytensor.cpp [hunk 3]
```diff
@@ -1711,6 +1700,144 @@ void pytensor_module(nb::module_& mod) {
...
+    // vs. the host-staging path (copy_device_to_host_tensor + host-side compose):
+    //   device DRAM --PCIe--> hugepage --memcpy--> host_staging --memcpy--> dest_torch (2 memcpys)
+    //
+    // Requires the dest_torch tensor to be sized for the device tensor's logical
+    // shape and dtype, contiguous in memory. For multi-shard meshes, each shard
+    // writes to its row-major slice (shards along dim 0 by convention).
+    mod.def(
+        "copy_device_to_torch",
+        [](const ttnn::Tensor& device_tensor,
+           const nb::ndarray<nb::pytorch>& dest,
```
Comment (nsextonTT) [inline ttnn/cpp/ttnn-nanobind/pytensor.cpp:1732]: The const& doesn't make much sense since you're populating it. It just happens to work since its holding a pointer under the hood. ndarray is just header information, so by-value would be more appropriate.

**Review by nsextonTT — APPROVED** [review]

File: ttnn/cpp/ttnn-nanobind/pytensor.cpp (hunk 3 above)
Comment (gtobarTT) [reply ttnn/cpp/ttnn-nanobind/pytensor.cpp:1732]: I appreciate the review
